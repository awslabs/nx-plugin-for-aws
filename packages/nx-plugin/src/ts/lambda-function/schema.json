{
  "$schema": "https://json-schema.org/schema",
  "$id": "TsLambdaFunction",
  "title": "Add a TypeScript Lambda Function to a project",
  "description": "Add a TypeScript Lambda Function to a project",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The project to add the lambda function to",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Select the project to add the lambda function to",
      "x-dropdown": "projects"
    },
    "functionName": {
      "type": "string",
      "description": "The name of the function to add",
      "x-priority": "important",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like your Lambda Function to have? i.e: MyFunction"
    },
    "functionPath": {
      "type": "string",
      "description": "The path within the project source directory to add the function to",
      "x-prompt": "Enter the path within the project source directory to add the function to (e.g. 'lambda-functions/')"
    },
    "handlerType": {
      "type": "string",
      "description": "The AWS Lambda handler type to use for the lambda function",
      "enum": [
        "APIGatewayProxyHandler",
        "APIGatewayProxyHandlerV2",
        "S3Handler",
        "SQSHandler",
        "SNSHandler",
        "DynamoDBStreamHandler",
        "KinesisStreamHandler",
        "CloudWatchLogsHandler",
        "ALBHandler",
        "LambdaFunctionURLHandler",
        "APIGatewayAuthorizerHandler",
        "S3BatchHandler",
        "SESHandler",
        "ScheduledHandler",
        "EventBridgeHandler",
        "Handler"
      ],
      "default": "APIGatewayProxyHandler",
      "x-prompt": "Select the AWS Lambda handler type for your function"
    }
  },
  "required": ["project", "functionName", "handlerType"]
}