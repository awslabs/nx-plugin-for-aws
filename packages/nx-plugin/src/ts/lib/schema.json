{
  "$schema": "https://json-schema.org/schema",
  "$id": "TsLib",
  "title": "Create a TypeScript Library",
  "description": "Create a TypeScript Library.",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Library name.",
      "x-priority": "important"
    },
    "directory": {
      "type": "string",
      "description": "Parent directory where the library is placed.",
      "default": "packages",
      "x-prompt": "Which directory do you want to create the library in?",
      "x-priority": "important"
    },
    "subDirectory": {
      "type": "string",
      "description": "The sub directory the lib is placed in. By default this is the library name.",
      "x-prompt": "Which sub directory do you want to create the library in? (By default this is the library name)"
    }
  },
  "required": ["name"]
}
