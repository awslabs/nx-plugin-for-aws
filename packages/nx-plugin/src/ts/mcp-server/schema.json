{
  "$schema": "https://json-schema.org/schema",
  "$id": "ts#mcp-server",
  "title": "ts#mcp-server",
  "description": "Generate a TypeScript Model Context Protocol (MCP) server for providing context to Large Language Models",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "MCP server project name.",
      "x-priority": "important"
    },
    "directory": {
      "type": "string",
      "description": "Parent directory where the MCP server is placed.",
      "default": "packages",
      "x-prompt": "Which directory do you want to create the MCP server in?",
      "x-priority": "important"
    },
    "subDirectory": {
      "type": "string",
      "description": "The sub directory the MCP server is placed in. By default this is the MCP server project name.",
      "x-prompt": "Which sub directory do you want to create the MCP server in? (By default this is the MCP server project name)"
    }
  },
  "required": ["name"]
}
