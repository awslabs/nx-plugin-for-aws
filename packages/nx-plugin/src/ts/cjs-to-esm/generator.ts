/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree, getProjects } from '@nx/devkit';
import { CjsToEsmGeneratorSchema } from './schema';
import { tsquery } from '@phenomnomnominal/tsquery';
import * as ts from 'typescript';
import { minimatch } from 'minimatch';
// Replaces "module.exports = value" with "export default value"
const replaceModuleExportsWithExportDefault = (
  sourceFile: ts.SourceFile
): ts.SourceFile => {
  const transformer =
    <T extends ts.Node>(context: ts.TransformationContext) =>
    (rootNode: T) => {
      function visit(node: ts.Node): ts.Node {
        if (ts.isExpressionStatement(node)) {
          const expr = node.expression;
          if (
            ts.isBinaryExpression(expr) &&
            ts.isPropertyAccessExpression(expr.left) &&
            ts.isIdentifier(expr.left.expression) &&
            expr.left.expression.text === 'module' &&
            ts.isIdentifier(expr.left.name) &&
            expr.left.name.text === 'exports' &&
            expr.operatorToken.kind === ts.SyntaxKind.EqualsToken
          ) {
            // Create an export default statement
            return context.factory.createExportAssignment(
              undefined,
              undefined,
              expr.right
            );
          }
        }
        return ts.visitEachChild(node, visit, context);
      }
      return ts.visitNode(rootNode, visit);
    };
  const result = ts.transform(sourceFile, [transformer]);
  return result.transformed[0] as ts.SourceFile;
};
// Replaces "const x = require('y')" with "import x from 'y'"
const replaceRequires = (sourceFile: ts.SourceFile): ts.SourceFile => {
  const transformer =
    <T extends ts.Node>(context: ts.TransformationContext) =>
    (rootNode: T) => {
      function visit(node: ts.Node): ts.Node {
        if (ts.isVariableStatement(node)) {
          const declaration = node.declarationList.declarations[0];
          if (declaration && ts.isVariableDeclaration(declaration)) {
            const initializer = declaration.initializer;
            if (
              initializer &&
              ts.isCallExpression(initializer) &&
              ts.isIdentifier(initializer.expression) &&
              initializer.expression.text === 'require' &&
              initializer.arguments.length === 1
            ) {
              const modulePath = initializer.arguments[0];
              if (ts.isStringLiteral(modulePath)) {
                // Handle destructuring pattern
                if (ts.isObjectBindingPattern(declaration.name)) {
                  const importSpecifiers = declaration.name.elements
                    .map((element) => {
                      if (
                        ts.isBindingElement(element) &&
                        ts.isIdentifier(element.name)
                      ) {
                        return context.factory.createImportSpecifier(
                          false,
                          undefined,
                          element.name
                        );
                      }
                      return undefined;
                    })
                    .filter(
                      (spec): spec is ts.ImportSpecifier => spec !== undefined
                    );
                  return context.factory.createImportDeclaration(
                    undefined,
                    context.factory.createImportClause(
                      false,
                      undefined,
                      context.factory.createNamedImports(importSpecifiers)
                    ),
                    modulePath,
                    undefined
                  );
                }
                // Handle simple require
                if (ts.isIdentifier(declaration.name)) {
                  return context.factory.createImportDeclaration(
                    undefined,
                    context.factory.createImportClause(
                      false,
                      declaration.name,
                      undefined
                    ),
                    modulePath,
                    undefined
                  );
                }
              }
            }
          }
        }
        return ts.visitEachChild(node, visit, context);
      }
      return ts.visitNode(rootNode, visit);
    };
  const result = ts.transform(sourceFile, [transformer]);
  return result.transformed[0] as ts.SourceFile;
};
// Replaces requires not directly assigned to variables eg
// "const x = require('y').foo()" with
// import _import0 from 'y'
// const x = _import0.foo()
const replaceInlineRequires = (sourceFile: ts.SourceFile): ts.SourceFile => {
  let importCounter = 0;
  const imports: ts.ImportDeclaration[] = [];
  const importMap = new Map<string, string>();
  const transformer =
    <T extends ts.Node>(context: ts.TransformationContext) =>
    (rootNode: T) => {
      function visit(node: ts.Node): ts.Node {
        if (
          ts.isCallExpression(node) &&
          ts.isIdentifier(node.expression) &&
          node.expression.text === 'require' &&
          node.arguments.length === 1 &&
          ts.isStringLiteral(node.arguments[0])
        ) {
          const modulePath = node.arguments[0];
          const modulePathText = modulePath.text;
          // Check if we already have an import for this module
          if (!importMap.has(modulePathText)) {
            const importName = `_import${importCounter++}`;
            importMap.set(modulePathText, importName);
            // Create import declaration
            imports.push(
              context.factory.createImportDeclaration(
                undefined,
                context.factory.createImportClause(
                  false,
                  context.factory.createIdentifier(importName),
                  undefined
                ),
                modulePath,
                undefined
              )
            );
          }
          // Replace require() with the import reference
          return context.factory.createIdentifier(
            importMap.get(modulePathText)
          );
        }
        return ts.visitEachChild(node, visit, context);
      }
      // First pass to collect all requires and create imports
      const visited = ts.visitNode(rootNode, visit);
      // If we found any inline requires, prepend the imports to the source file
      if (imports.length > 0 && ts.isSourceFile(visited)) {
        return context.factory.updateSourceFile(visited, [
          ...imports,
          ...visited.statements,
        ]);
      }
      return visited;
    };
  const result = ts.transform(sourceFile, [transformer]);
  return result.transformed[0] as ts.SourceFile;
};
export function cjsToEsm(
  tree: Tree,
  relativePathInTree: string,
  options?: {
    include?: string[];
    exclude?: string[];
  }
) {
  const includePatterns = options?.include ?? ['**/*.{js,ts}'];
  const excludePatterns =
    options?.exclude ??
    ['node_modules', 'dist', 'build', 'tmp', '.nx'].flatMap((p) => [
      p,
      `**/${p}/**/*`,
      `${p}/**/*`,
    ]);
  // Process each TypeScript/JavaScript file in the project
  const processFile = (filePath: string) => {
    const content = tree.read(filePath, 'utf-8');
    if (!content) return;
    // Parse the source file
    const originalSourceFile = tsquery.ast(content);
    let sourceFile = originalSourceFile;
    sourceFile = replaceModuleExportsWithExportDefault(sourceFile);
    sourceFile = replaceRequires(sourceFile);
    sourceFile = replaceInlineRequires(sourceFile);
    const printer = ts.createPrinter({
      removeComments: false,
      newLine: ts.NewLineKind.LineFeed,
    });
    tree.write(
      filePath,
      printer.printNode(ts.EmitHint.Unspecified, sourceFile, originalSourceFile)
    );
  };
  // Process all files in the given directory (recursively)
  const processDirectory = (dirPath: string) => {
    const entries = tree.children(dirPath);
    entries.forEach((entry) => {
      const fullPath = `${dirPath}/${entry}`;
      const relativePath = fullPath.replace(`${relativePathInTree}/`, '');
      // Return early for exclude to ensure we don't traverse node_modules etc!
      const shouldExclude = excludePatterns.some((pattern) =>
        minimatch(relativePath, pattern)
      );
      if (shouldExclude) {
        return;
      }
      if (tree.isFile(fullPath)) {
        const shouldInclude = includePatterns.some((pattern) =>
          minimatch(relativePath, pattern)
        );
        if (shouldInclude) {
          processFile(fullPath);
        }
      } else {
        processDirectory(fullPath);
      }
    });
  };
  // Start processing from the given directories
  processDirectory(relativePathInTree);
}
export async function cjsToEsmGenerator(
  tree: Tree,
  options: CjsToEsmGeneratorSchema
) {
  // Get the project configuration
  const projects = getProjects(tree);
  const project = projects.get(options.project);
  if (!project) {
    throw new Error(`Project ${options.project} not found`);
  }
  cjsToEsm(tree, project.root, {
    include: options.include ? [options.include] : undefined,
    exclude: options.exclude ? [options.exclude] : undefined,
  });
}
export default cjsToEsmGenerator;
