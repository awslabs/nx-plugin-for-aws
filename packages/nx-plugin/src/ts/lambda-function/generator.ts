/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  generateFiles,
  GeneratorCallback,
  installPackagesTask,
  joinPathFragments,
  OverwriteStrategy,
  ProjectConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { TsLambdaFunctionGeneratorSchema } from './schema';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from '../../utils/shared-constructs-constants';
import { toClassName, toKebabCase, pascalCase } from '../../utils/names';
import { addStarExport } from '../../utils/ast';
import { formatFilesInSubtree } from '../../utils/format';
import { sortObjectKeys } from '../../utils/object';
import {
  NxGeneratorInfo,
  getGeneratorInfo,
  readProjectConfigurationUnqualified,
} from '../../utils/nx';
import { addGeneratorMetricsIfApplicable } from '../../utils/metrics';
import { withVersions } from '../../utils/versions';
import camelCase from 'lodash.camelcase';
import { TS_HANDLER_RETURN_TYPES } from './io';
import { addEsbuildBundleTarget } from '../../utils/esbuild';

export const TS_LAMBDA_FUNCTION_GENERATOR_INFO: NxGeneratorInfo =
  getGeneratorInfo(__filename);

/**
 * Generates a TypeScript Lambda Function to add to a TypeScript project
 */
export const tsLambdaFunctionGenerator = async (
  tree: Tree,
  schema: TsLambdaFunctionGeneratorSchema,
): Promise<GeneratorCallback> => {
  const projectConfig = readProjectConfigurationUnqualified(
    tree,
    schema.project,
  );

  const tsconfigPath = joinPathFragments(projectConfig.root, 'tsconfig.json');

  // Check if the project has a tsconfig.json file
  if (!tree.exists(tsconfigPath)) {
    throw new Error(
      `This generator does not support selected project ${schema.project}. The project must be a typescript project (ie contain a tsconfig.json)`,
    );
  }

  if (!projectConfig.sourceRoot) {
    throw new Error(
      `This project does not have a source root. Please add a source root to the project configuration before running this generator.`,
    );
  }

  const dir = projectConfig.root;
  const projectNameWithOutScope = projectConfig.name.split('/').pop();
  const projectNameKebabCase = toKebabCase(projectNameWithOutScope);
  const functionNameKebabCase = toKebabCase(schema.functionName);

  const constructFunctionNameKebabCase = `${projectNameKebabCase}-${functionNameKebabCase}`;
  const constructFunctionClassName = toClassName(
    constructFunctionNameKebabCase,
  );
  const lambdaFunctionCamelCase = camelCase(schema.functionName);
  const lambdaFunctionClassName = pascalCase(schema.functionName);
  const lambdaFunctionKebabCase = toKebabCase(schema.functionName);

  const functionPath = joinPathFragments(
    projectConfig.sourceRoot,
    schema.functionPath ?? '',
    `${lambdaFunctionKebabCase}.ts`,
  );

  // Check that the project does not already have a lambda handler
  if (tree.exists(functionPath)) {
    throw new Error(
      `This project already has a lambda function with the name ${functionNameKebabCase}. Please remove the lambda function before running this generator or use a different name.`,
    );
  }

  await sharedConstructsGenerator(tree);

  // Add bundle-<name> target for this specific lambda function
  const bundleTargetName = `bundle-${lambdaFunctionKebabCase}`;

  const enhancedOptions = {
    ...schema,
    dir,
    constructFunctionClassName,
    constructFunctionNameKebabCase,
    lambdaFunctionCamelCase,
    lambdaFunctionClassName,
    lambdaFunctionKebabCase,
    bundleTargetName,
    returnType: TS_HANDLER_RETURN_TYPES[schema.eventSource],
  };

  // Add a bundle target for the function
  addEsbuildBundleTarget(projectConfig, {
    bundleTargetName,
    targetFilePath: functionPath,
    extraEsbuildArgs: '--external:@aws-sdk/*',
  });

  projectConfig.targets = sortObjectKeys(projectConfig.targets);
  updateProjectConfiguration(tree, projectConfig.name, projectConfig);

  // Generate the lambda handler file
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'handler'),
    joinPathFragments(projectConfig.sourceRoot, schema.functionPath ?? ''),
    enhancedOptions,
    { overwriteStrategy: OverwriteStrategy.KeepExisting },
  );

  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', SHARED_CONSTRUCTS_DIR, 'src', 'app'),
    joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'src', 'app'),
    enhancedOptions,
    { overwriteStrategy: OverwriteStrategy.KeepExisting },
  );

  addStarExport(
    tree,
    joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'index.ts',
    ),
    './lambda-functions/index.js',
  );
  addStarExport(
    tree,
    joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'lambda-functions',
      'index.ts',
    ),
    `./${constructFunctionNameKebabCase}.js`,
  );

  // Ensure common constructs builds after our lambda function project
  updateJson(
    tree,
    joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
    (config: ProjectConfiguration) => {
      if (!config.targets) {
        config.targets = {};
      }
      if (!config.targets.build) {
        config.targets.build = {};
      }
      config.targets.build.dependsOn = [
        ...(config.targets.build.dependsOn ?? []).filter(
          (t) => t !== `${projectConfig.name}:build`,
        ),
        `${projectConfig.name}:build`,
      ];
      return config;
    },
  );

  addDependenciesToPackageJson(
    tree,
    withVersions([
      '@aws-lambda-powertools/tracer',
      '@aws-lambda-powertools/logger',
      '@aws-lambda-powertools/metrics',
      '@aws-lambda-powertools/parser',
      '@middy/core',
      'zod',
    ]),
    withVersions(['esbuild', '@types/aws-lambda']),
  );

  await addGeneratorMetricsIfApplicable(tree, [
    TS_LAMBDA_FUNCTION_GENERATOR_INFO,
  ]);

  await formatFilesInSubtree(tree);

  return () => {
    installPackagesTask(tree);
  };
};

export default tsLambdaFunctionGenerator;
