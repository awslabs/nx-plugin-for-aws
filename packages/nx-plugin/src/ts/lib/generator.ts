/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  GeneratorCallback,
  NxJsonConfiguration,
  OverwriteStrategy,
  Tree,
  generateFiles,
  getPackageManagerCommand,
  installPackagesTask,
  joinPathFragments,
  readProjectConfiguration,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { TsLibGeneratorSchema } from './schema';
import { libraryGenerator } from '@nx/js';
import { getNpmScopePrefix } from '../../utils/npm-scope';
import { configureTsProject } from './ts-project-utils';
import { toKebabCase } from '../../utils/names';
import { relative } from 'path';
import { formatFilesInSubtree } from '../../utils/format';
import { sortObjectKeys } from '../../utils/nx';
export interface TsLibDetails {
  /**
   * Full package name including scope (eg @foo/bar)
   */
  readonly fullyQualifiedName: string;
  /**
   * Directory of the library relative to the root
   */
  readonly dir: string;
}
/**
 * Returns details about the TS library to be created
 */
export const getTsLibDetails = (
  tree: Tree,
  schema: TsLibGeneratorSchema,
): TsLibDetails => {
  const scope = getNpmScopePrefix(tree);
  const normalizedName = toKebabCase(schema.name);
  const fullyQualifiedName = `${scope}${normalizedName}`;
  const dir = joinPathFragments(
    toKebabCase(schema.directory) ?? '.',
    toKebabCase(schema.subDirectory) ?? normalizedName,
  );
  return { dir, fullyQualifiedName };
};
/**
 * Generates a typescript library
 */
export const tsLibGenerator = async (
  tree: Tree,
  schema: TsLibGeneratorSchema,
): Promise<GeneratorCallback> => {
  const { fullyQualifiedName, dir } = getTsLibDetails(tree, schema);
  await libraryGenerator(tree, {
    ...schema,
    name: toKebabCase(schema.name),
    directory: dir,
    skipPackageJson: true,
    bundler: 'tsc', // TODO: consider supporting others
    linter: 'eslint',
    unitTestRunner: 'vitest',
  });
  // Replace with simpler sample source code
  tree.delete(joinPathFragments(dir, 'src'));
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    joinPathFragments(dir),
    {
      fullyQualifiedName,
      pkgMgrCmd: getPackageManagerCommand().exec,
    },
    {
      overwriteStrategy: OverwriteStrategy.KeepExisting,
    },
  );
  configureTsProject(tree, {
    dir,
    fullyQualifiedName,
  });

  const projectConfiguration = readProjectConfiguration(
    tree,
    fullyQualifiedName,
  );
  const targets = projectConfiguration.targets;

  targets['compile'] = {
    executor: 'nx:run-commands',
    outputs: [`{workspaceRoot}/dist/${dir}/tsc`],
    options: {
      command: 'tsc --build tsconfig.lib.json',
      cwd: '{projectRoot}',
    },
  };
  targets['build'] = {
    dependsOn: ['lint', 'compile', 'test'],
  };
  targets['test'] = {
    executor: '@nx/vite:test',
    outputs: ['{options.reportsDirectory}'],
    options: {
      reportsDirectory: joinPathFragments(
        relative(joinPathFragments(tree.root, dir), tree.root),
        'coverage',
        dir,
      ),
    },
  };
  delete targets['lint'];
  projectConfiguration.targets = sortObjectKeys(targets);

  updateProjectConfiguration(tree, fullyQualifiedName, projectConfiguration);

  updateJson(tree, 'nx.json', (nxJson: NxJsonConfiguration) => {
    nxJson.plugins = nxJson.plugins.map((p) => {
      if (
        typeof p !== 'string' &&
        p.plugin === '@nx/js/typescript' &&
        p.options?.['build']
      ) {
        p.options['build'].targetName = 'compile';
      }
      return p;
    });
    return nxJson;
  });

  await formatFilesInSubtree(tree);

  return () => {
    if (!schema.skipInstall) {
      installPackagesTask(tree);
    }
  };
};
export default tsLibGenerator;
