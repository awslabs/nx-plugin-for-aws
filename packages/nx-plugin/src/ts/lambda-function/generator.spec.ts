import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { tsLambdaFunctionGenerator } from './generator';
import { TsLambdaFunctionSchema } from './schema';

describe('ts lambda-function generator', () => {
  it('should generate a basic Lambda function with APIGatewayProxyHandler', async () => {
    const tree = createTreeUsingTsSolutionSetup();

    const options: TsLambdaFunctionSchema = {
      project: 'test-project',
      functionName: 'MyFunction',
      handlerType: 'APIGatewayProxyHandler',
    };

    await tsLambdaFunctionGenerator(tree, options);

    // Check that the handler file was created
    expect(tree.exists('packages/test-project/src/lambda-functions/my-function.ts')).toBe(true);
    
    // Check that the test file was created
    expect(tree.exists('packages/test-project/src/tests/my-function.spec.ts')).toBe(true);
    
    // Check that the CDK construct was created
    expect(tree.exists('packages/common/constructs/src/app/lambda-functions/test-project-my-function.ts')).toBe(true);

    // Check handler content
    const handlerContent = tree.read('packages/test-project/src/lambda-functions/my-function.ts', 'utf-8');
    expect(handlerContent).toContain('import { APIGatewayProxyHandler } from \'aws-lambda\'');
    expect(handlerContent).toContain('export const handler: APIGatewayProxyHandler');
    expect(handlerContent).toContain('statusCode: 200');

    // Check test content
    const testContent = tree.read('packages/test-project/src/tests/my-function.spec.ts', 'utf-8');
    expect(testContent).toContain('describe(\'MyFunction Lambda Handler\'');
    expect(testContent).toContain('should handle API Gateway proxy event');

    // Check CDK construct content
    const constructContent = tree.read('packages/common/constructs/src/app/lambda-functions/test-project-my-function.ts', 'utf-8');
    expect(constructContent).toContain('export class TestProjectMyFunction extends Function');
    expect(constructContent).toContain('handler: \'my-function.handler\'');
  });

  it('should generate a Lambda function with S3Handler', async () => {
    const tree = createTreeUsingTsSolutionSetup();

    const options: TsLambdaFunctionSchema = {
      project: 'test-project',
      functionName: 'S3Processor',
      handlerType: 'S3Handler',
      functionPath: 'handlers',
    };

    await tsLambdaFunctionGenerator(tree, options);

    // Check that the handler file was created in the specified path
    expect(tree.exists('packages/test-project/src/handlers/s3-processor.ts')).toBe(true);

    // Check handler content
    const handlerContent = tree.read('packages/test-project/src/handlers/s3-processor.ts', 'utf-8');
    expect(handlerContent).toContain('import { S3Handler } from \'aws-lambda\'');
    expect(handlerContent).toContain('export const handler: S3Handler');
    expect(handlerContent).toContain('for (const record of event.Records)');
  });

  it('should generate a Lambda function with any handler type', async () => {
    const tree = createTreeUsingTsSolutionSetup();

    const options: TsLambdaFunctionSchema = {
      project: 'test-project',
      functionName: 'GenericFunction',
      handlerType: 'any',
    };

    await tsLambdaFunctionGenerator(tree, options);

    // Check handler content
    const handlerContent = tree.read('packages/test-project/src/lambda-functions/generic-function.ts', 'utf-8');
    expect(handlerContent).toContain('import { Context } from \'aws-lambda\'');
    expect(handlerContent).toContain('export const handler = async (event: any, context: Context)');
    expect(handlerContent).not.toContain('@aws-lambda-powertools');
  });

  it('should update project configuration with bundle target', async () => {
    const tree = createTreeUsingTsSolutionSetup();

    const options: TsLambdaFunctionSchema = {
      project: 'test-project',
      functionName: 'TestFunction',
    };

    await tsLambdaFunctionGenerator(tree, options);

    // Check that project.json was updated
    const projectJson = JSON.parse(tree.read('packages/test-project/project.json', 'utf-8') || '{}');
    expect(projectJson.targets?.bundle).toBeDefined();
    expect(projectJson.targets.bundle.executor).toBe('@nx/esbuild:esbuild');
    expect(projectJson.targets.bundle.options.main).toBe('packages/test-project/src/lambda-functions/test-function.ts');
  });

  it('should add dependencies to package.json', async () => {
    const tree = createTreeUsingTsSolutionSetup();

    const options: TsLambdaFunctionSchema = {
      project: 'test-project',
      functionName: 'TestFunction',
      handlerType: 'EventBridgeHandler',
    };

    await tsLambdaFunctionGenerator(tree, options);

    // Check that package.json was updated
    const packageJson = JSON.parse(tree.read('packages/test-project/package.json', 'utf-8') || '{}');
    expect(packageJson.devDependencies?.['@types/aws-lambda']).toBeDefined();
    expect(packageJson.dependencies?.['@aws-lambda-powertools/logger']).toBeDefined();
    expect(packageJson.dependencies?.['@aws-lambda-powertools/metrics']).toBeDefined();
    expect(packageJson.dependencies?.['@aws-lambda-powertools/tracer']).toBeDefined();
  });
});