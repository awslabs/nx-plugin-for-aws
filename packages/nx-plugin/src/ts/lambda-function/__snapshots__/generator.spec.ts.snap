// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ts-lambda-function generator > should create CDK construct > cdk-construct.ts 1`] = `
"import { Construct } from 'constructs';
import * as url from 'url';
import { Code, Function, Runtime, Tracing } from 'aws-cdk-lib/aws-lambda';
import { Duration } from 'aws-cdk-lib';

export class TestProjectTestFunction extends Function {
  constructor(scope: Construct, id: string) {
    super(scope, id, {
      timeout: Duration.seconds(30),
      runtime: Runtime.NODEJS_LATEST,
      handler: 'index.handler',
      code: Code.fromAsset(
        url.fileURLToPath(
          new URL(
            '../../../../../../dist/packages/test-project/bundle-test-function',
            import.meta.url,
          ),
        ),
      ),
      tracing: Tracing.ACTIVE,
      environment: {
        AWS_CONNECTION_REUSE_ENABLED: '1',
      },
    });
  }
}
"
`;

exports[`ts-lambda-function generator > should create lambda function file with Any event source > lambda-handler-any.ts 1`] = `
"import middy from '@middy/core';
import { Tracer } from '@aws-lambda-powertools/tracer';
import { captureLambdaHandler } from '@aws-lambda-powertools/tracer/middleware';
import { injectLambdaContext } from '@aws-lambda-powertools/logger/middleware';
import { Logger } from '@aws-lambda-powertools/logger';
import { Metrics } from '@aws-lambda-powertools/metrics';
import { logMetrics } from '@aws-lambda-powertools/metrics/middleware';

process.env.POWERTOOLS_METRICS_NAMESPACE = 'TestFunction';
process.env.POWERTOOLS_SERVICE_NAME = 'TestFunction';

const tracer = new Tracer();
const logger = new Logger();
const metrics = new Metrics();

export const testFunction = async (event: any): Promise<any> => {
  logger.info('Received event', event);

  // TODO: implement
};

export const handler = middy()
  .use(captureLambdaHandler(tracer))
  .use(injectLambdaContext(logger))
  .use(logMetrics(metrics))
  .handler(testFunction);
"
`;

exports[`ts-lambda-function generator > should create lambda function file with EventBridge schema > lambda-handler-eventbridge.ts 1`] = `
"import { parser } from '@aws-lambda-powertools/parser/middleware';
import { EventBridgeSchema } from '@aws-lambda-powertools/parser/schemas';
import { z } from 'zod';
import middy from '@middy/core';
import { Tracer } from '@aws-lambda-powertools/tracer';
import { captureLambdaHandler } from '@aws-lambda-powertools/tracer/middleware';
import { injectLambdaContext } from '@aws-lambda-powertools/logger/middleware';
import { Logger } from '@aws-lambda-powertools/logger';
import { Metrics } from '@aws-lambda-powertools/metrics';
import { logMetrics } from '@aws-lambda-powertools/metrics/middleware';

process.env.POWERTOOLS_METRICS_NAMESPACE = 'TestFunction';
process.env.POWERTOOLS_SERVICE_NAME = 'TestFunction';

const tracer = new Tracer();
const logger = new Logger();
const metrics = new Metrics();

export const testFunction = async (
  event: z.infer<typeof EventBridgeSchema>,
): Promise<void> => {
  logger.info('Received event', event);

  // TODO: implement
};

export const handler = middy()
  .use(captureLambdaHandler(tracer))
  .use(injectLambdaContext(logger))
  .use(logMetrics(metrics))
  .use(parser({ schema: EventBridgeSchema }))
  .handler(testFunction);
"
`;

exports[`ts-lambda-function generator > should create lambda function file with SQS schema > lambda-handler-sqs.ts 1`] = `
"import { parser } from '@aws-lambda-powertools/parser/middleware';
import { SqsSchema } from '@aws-lambda-powertools/parser/schemas';
import { z } from 'zod';
import middy from '@middy/core';
import { Tracer } from '@aws-lambda-powertools/tracer';
import { captureLambdaHandler } from '@aws-lambda-powertools/tracer/middleware';
import { injectLambdaContext } from '@aws-lambda-powertools/logger/middleware';
import { Logger } from '@aws-lambda-powertools/logger';
import { Metrics } from '@aws-lambda-powertools/metrics';
import { logMetrics } from '@aws-lambda-powertools/metrics/middleware';
import type { SQSBatchResponse } from 'aws-lambda';

process.env.POWERTOOLS_METRICS_NAMESPACE = 'TestFunction';
process.env.POWERTOOLS_SERVICE_NAME = 'TestFunction';

const tracer = new Tracer();
const logger = new Logger();
const metrics = new Metrics();

export const testFunction = async (
  event: z.infer<typeof SqsSchema>,
): Promise<SQSBatchResponse | void> => {
  logger.info('Received event', event);

  // TODO: implement
};

export const handler = middy()
  .use(captureLambdaHandler(tracer))
  .use(injectLambdaContext(logger))
  .use(logMetrics(metrics))
  .use(parser({ schema: SqsSchema }))
  .handler(testFunction);
"
`;
